### PR TITLE
MDEV-39520: Add MySQL 8.0 extended syntax for `REGEXP_INSTR`

### PR DESCRIPTION
**Pull request by MohamedM216**
## Reference Issue 
[MDEV-39520](https://jira.mariadb.org/browse/MDEV-39520)

## Description
This PR is one of a series of PRs that are supposed to add MySQL 8.0-compatible extended syntax for all REGEXP functions

## Test
regexp_instr_mysql8.test

### CODE REVIEW

File: sql/item_cmpfunc.cc [hunk 1]
```diff
@@ -6521,21 +6567,145 @@ Item_func_regexp_instr::fix_length_and_dec(THD *thd)
     return TRUE;
 
   re.init(cmp_collation.collation, 0);
+
+  if (arg_count > 5 && args[5]->const_item())
```
Comment (gkodinov) [inline sql/item_cmpfunc.cc]: Do you *really* need to be doing this here? These flags aren't used until compile is called. And it's not called during compilation.

File: sql/item_cmpfunc.h [hunk 2]
```diff
@@ -3274,6 +3274,7 @@ class Regexp_processor_pcre
   void set_const(bool arg) { m_is_const= arg; }
   CHARSET_INFO * library_charset() const { return m_library_charset; }
   void unset_flag(int flag) { m_library_flags&= ~flag; }
+  void set_library_flags(int flags) { m_library_flags= flags; }
```
Comment (gkodinov) [inline sql/item_cmpfunc.h]: Why don't you do the parsing inside this function? And merge parse_match_type_flags into it? There's no call sequence in your pull request where the two are not called one after another.

**Review by gkodinov — CHANGES_REQUESTED** [review]
Thank you for your contribution! This is a preliminary review.

First of all: this is implementing just one sub-task of what https://jira.mariadb.org/browse/MDEV-39106 calls for. Either do all of it, as scoped in the jira, in a single go. Or create Jira sub-task(s) into the above jira and target these instead. 

There's also some suggestions below.

File: sql/item_cmpfunc.cc (hunk 1 above)
Comment (MohamedM216) [reply sql/item_cmpfunc.cc]: Yes, to handle constant pattern case, while it's not important for the non-constant pattern (we handle non-constant in `val_int()`). The non-constant pattern leads to set `m_is_const = false` in `fix_owner()` which doesn't block `recompile()`, so if we remove this block of code from `fix_length_and_dec()`, non-constant case results won't be affected. But we still have another case: constant pattern. It sets `m_is_const = true` in `fix_owner()` and that blocks `recompile()`
```cpp
bool recompile(Item *item)
  {
    return !m_is_const && compile(item, false);
  }
```
so no matter whether or not we update `m_library_flags`, it won't be affected and won't affect the final result as well, as `compile()` will always return the cached value for every row while executing this statement. And even if we tried to call `compile()` directly in `val_int()` to bypass `recompile()`, it wouldn't help either as `compile()` has its own cache guard that skips recompilation when the pattern string hasn't changed, which it hasn't for a constant pattern. And that's why the test fails with these statements if we try to remove this block of code from `fix_length_and_dec()`:
```bash
@@ -1,9 +1,9 @@
 SELECT REGEXP_INSTR('Abba', 'ABBA', 1, 1, 0, 'c');
 REGEXP_INSTR('Abba', 'ABBA', 1, 1, 0, 'c')
-0
+1
 SELECT REGEXP_INSTR('Abba', 'ABBA', 1, 1, 0, 'ic');
 REGEXP_INSTR('Abba', 'ABBA', 1, 1, 0, 'ic')
-0
+1
 SELECT REGEXP_INSTR('aabbaabb', '(a+)(b+)\\1', 1, 1);
 REGEXP_INSTR('aabbaabb', '(a+)(b+)\\1', 1, 1)
 1
@@ -12,4 +12,4 @@
 14
 SELECT REGEXP_INSTR('Test-abc-abc-abc', 'AB', 1, 3, 0, 'c');
 REGEXP_INSTR('Test-abc-abc-abc', 'AB', 1, 3, 0, 'c')
-0
+14

Result length mismatch
```

Comment (gkodinov) [reply sql/item_cmpfunc.cc]: Thanks for the explanation. Makes sense. This also brings me to one question: how will it behave in cases as follows:

```
CREATE table foo(cond varchar(50));
INSERT INTO foo VALUES ('c', 'ic');
SELECT regexp_instr(Abba', 'ABBA', 1, 1, 0, cond) FROM foo.

```
As in: the code is not a constant and the rest of the condition is.

I believe that you will need to consider the whole thing non-constant if any of the compile- affecting elements are non-constant. 

Would you please add a test like this?

File: sql/item_cmpfunc.h (hunk 2 above)
Comment (MohamedM216) [reply sql/item_cmpfunc.h]: Addressed here https://github.com/MariaDB/server/compare/ef7d499f0aab4ce4ecbbe27db5d9132d8f50bd9b..d0e479a924030d3caf71cea37a5d3f0e2d00cf0b

File: sql/item_cmpfunc.cc (hunk 1 above)
Comment (MohamedM216) [reply sql/item_cmpfunc.cc]: >how will it behave in cases as follows

It currently returns 
```
1
1
```
which confirms the case you described. So to fix this (non-constant `match_type` with constant `pattern`) I invalidated the pattern cache so `compile()` is not skipped when flags change and the result now is
```
0
0
```
Changes addressed here https://github.com/MariaDB/server/compare/d0e479a924030d3caf71cea37a5d3f0e2d00cf0b..2634104f490c5c8d40d7463506c0b68ec74ff0a8

File: mysql-test/main/regexp_instr_mysql8.test [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+# case-sensitive vs collation default
+SELECT REGEXP_INSTR('Test-abc-abc-abc', 'AB', 1, 3, 0);
+SELECT REGEXP_INSTR('Test-abc-abc-abc', 'AB', 1, 3, 0, 'c');
+
+--echo # Non-constant match_type with constant pattern
+CREATE TABLE foo (cond VARCHAR(50));
+INSERT INTO foo VALUES ('c'), ('ic');
+SELECT REGEXP_INSTR('Abba', 'ABBA', 1, 1, 0, cond) FROM foo;
+DROP TABLE foo;
```
Comment (gkodinov) [inline mysql-test/main/regexp_instr_mysql8.test]: please always terminate the last line in any file with LF. We also customarily end the .test files with `--echo End of 13.0 tests`

File: sql/item_cmpfunc.cc [hunk 4]
```diff
@@ -6521,21 +6521,139 @@ Item_func_regexp_instr::fix_length_and_dec(THD *thd)
...
+    String *match_type_str= args[5]->val_str(&mt_tmp);
+    if ((null_value= args[5]->null_value))
+      return 0;
+
+    if (!args[5]->const_item())
+    {
+      re.init(cmp_collation.collation, 0);
+      if (re.parse_match_type_flags(match_type_str, cmp_collation))
+        return (null_value= true, 0);
+      if ((null_value= re.compile(args[1], false)))
```
Comment (gkodinov) [inline sql/item_cmpfunc.cc:6673]: Not exactly sure why you need to compile this unconditionally after you've just called recompile above. 
I'd suggest moving the whole thing before the re-compile call.

File: sql/item_cmpfunc.h [hunk 5]
```diff
@@ -17,6 +17,9 @@
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1335  USA */
 
 
+#define PCRE2_STATIC 1             /* Important on Windows */
+#include "pcre2.h"                 /* pcre2 header file */
```
Comment (gkodinov) [inline sql/item_cmpfunc.h]: I guess this is what causes windows builds to fail. Remove this from here and move the new inlined function into the .cc file: it already includes the right headers.

**Review by gkodinov — CHANGES_REQUESTED** [review]
Please fix the failing buildbot tests. Some improvement suggestions below too.

File: mysql-test/main/regexp_instr_mysql8.test (hunk 3 above)
Comment (MohamedM216) [reply mysql-test/main/regexp_instr_mysql8.test]: done

File: sql/item_cmpfunc.h (hunk 5 above)
Comment (MohamedM216) [reply sql/item_cmpfunc.h]: updated

**Comment by MohamedM216** [comment]
The current failing tests aren't related to my patch and they pass locally.

File: sql/item_cmpfunc.cc (hunk 4 above)
Comment (MohamedM216) [reply sql/item_cmpfunc.cc:6673]: `recompile()` cannot replace `compile()` here because `m_is_const=true` for constant patterns blocks it, making the direct `compile()` call in the non-const `match_type` block unavoidable. I suggest keeping it as it is. I'd appreciate your feedback.

Comment (gkodinov) [reply sql/item_cmpfunc.cc:6673]: OK, something for the final review I guess.

**Review by gkodinov — APPROVED** [review]
LGTM. Please stand by for the final review.

File: sql/item_cmpfunc.cc [hunk 6]
```diff
@@ -6514,28 +6514,199 @@ bool Item_func_regex::val_bool()
...
+
 bool
 Item_func_regexp_instr::fix_length_and_dec(THD *thd)
 {
   if (agg_arg_charsets_for_comparison(cmp_collation, args, 2))
     return TRUE;
 
   re.init(cmp_collation.collation, 0);
+
+  if (arg_count > 5 && args[5]->const_item())
```
Comment (sanja-byelkin) [inline sql/item_cmpfunc.cc]: const_item is not basic constant. prepare place holder (parameter '?') is a constant for example but changes, some other cases also possible if I remember correctly. So tests with changing parameter of prepared statement should be added.

File: sql/item_cmpfunc.cc [hunk 7]
```diff
@@ -6514,28 +6514,199 @@ bool Item_func_regex::val_bool()
...
+      return (null_value= true, 0);
+    }
+  }
+
+  // args[5]: match_type string
+  if (arg_count > 5)
+  {
+    char mt_buf[64];
+    String mt_tmp(mt_buf, sizeof(mt_buf), &my_charset_latin1);
+    String *match_type_str= args[5]->val_str(&mt_tmp);
```
Comment (sanja-byelkin) [inline sql/item_cmpfunc.cc]: If It was constant and flags was calculated on fix field what sens to repeat here again and again the same exercise?

File: sql/item_cmpfunc.cc [hunk 8]
```diff
@@ -6514,28 +6514,199 @@ bool Item_func_regex::val_bool()
...
+    if ((null_value= args[5]->null_value))
+      return 0;
+
+    if (!args[5]->const_item())
```
Comment (sanja-byelkin) [inline sql/item_cmpfunc.cc]: And here we are suppose to use old value but above we get string value of the Item what the sens in above evaluation?

Comment (sanja-byelkin) [reply sql/item_cmpfunc.cc]: Also constant flag can change try this with tables of 1 row length and it will fail (please add such test also). i.e. if we cached values on fix_field w should not detect by call of const_item during execution bacouse we have make_const calls, we have substitution and maybe I forgot about something else. So please use something else (maybe just boolean flag)

File: sql/item_cmpfunc.cc [hunk 9]
```diff
@@ -6514,28 +6514,199 @@ bool Item_func_regex::val_bool()
...
+      re.init(cmp_collation.collation, 0);
+      if (parse_match_type_flags(re, match_type_str, cmp_collation))
+        return (null_value= true, 0);
+      if ((null_value= re.compile(args[1], false)))
+        return 0;
+    }
+  }
+
+  char subject_buf[MAX_FIELD_WIDTH];
+  String subject_tmp(subject_buf, sizeof(subject_buf), &my_charset_bin);
```
Comment (sanja-byelkin) [inline sql/item_cmpfunc.cc:6679]: I am not specialist in regexp but I feel it can be not so efficient as previous code. So testing of performance should be done. @gkodinov  could you organise one. Also it is new feature which rewrite rexexp so after code review it have to go to a tester review as we do with all features.

**Review by sanja-byelkin — CHANGES_REQUESTED** [review]
Found implementation problem should be fixed. Also performance and functionality test should be made after code review done

File: sql/item_cmpfunc.cc (hunk 8 above)
Comment (MohamedM216) [reply sql/item_cmpfunc.cc]: >Also constant flag can change try this with tables of 1 row length and it will fail

You're right, it fails, but using the boolean trick to store `const_item()`'s value then reuse it fixes this bug.

The test is found at the end of the test file.

File: sql/item_cmpfunc.cc (hunk 6 above)
Comment (MohamedM216) [reply sql/item_cmpfunc.cc]: A test for this case is added now at the end of the test file.

**Comment by MohamedM216** [comment]
Thanks @sanja-byelkin for the detailed feedback!

**Comment by MohamedM216** [comment]
Hi @sanja-byelkin, just following up on this PR. I'd appreciate your feedback. Thanks!

**Comment by sanja-byelkin** [comment]
I buildbot above tests failing with memory leak around regexp

**Comment by MohamedM216** [comment]
@sanja-byelkin I believe these failures are unrelated to my PR; the `mariabackup.incremental_compressed` crash is an InnoDB redo log recovery assertion in `log0recv.cc` and `perfschema.socket_connect` is a networking test.

File: sql/item_cmpfunc.cc [hunk 10]
```diff
@@ -6516,28 +6516,218 @@ bool Item_func_regex::val_bool()
 }
 
 
+// false = OK, true = error
+static bool parse_match_type_flags(Regexp_processor_pcre &re,
```
Comment (gkodinov) [inline sql/item_cmpfunc.cc]: This is not just parsing stuff. it's also applying it to the re context. 
Either name the function apply_match_type() or make it just parse and return the types detected. 
I'd actually make it be an overloaded init() method for Regexp_processor_pcre(). This way you can avoid the awkward re.unset_flag()/re.set_flag() dance.

File: sql/item_cmpfunc.cc [hunk 11]
```diff
@@ -6516,28 +6516,218 @@ bool Item_func_regex::val_bool()
...
+    {
+      if (parse_match_type_flags(re, match_type_str, cmp_collation))
+        return TRUE;
+    }
+  }
   max_length= MY_INT32_NUM_DECIMAL_DIGITS; // See also Item_func_locate
   return re.fix_owner(this, args[0], args[1]);
 }
 
-
```
Comment (gkodinov) [inline sql/item_cmpfunc.cc:6530]: do not do white-space only changes please.

File: sql/item_cmpfunc.cc (hunk 9 above)
Comment (gkodinov) [reply sql/item_cmpfunc.cc:6679]: this is going to be part of the testing in general I guess.

**Review by gkodinov — APPROVED** [review]
LGTM. Please consider squashing the two commits before the end of the review. 

A stylistic change proposed below.

File: sql/item_cmpfunc.cc (hunk 11 above)
Comment (MohamedM216) [reply sql/item_cmpfunc.cc:6530]: Just by mistake :)

File: sql/item_cmpfunc.cc (hunk 10 above)
Comment (MohamedM216) [reply sql/item_cmpfunc.cc]: Thanks for the feedback!
I have removed the standalone helper function and implemented the overloaded `init(data_charset, match_type_str, collation)` method for `Regexp_processor_pcre`.

File: sql/item_cmpfunc.cc [hunk 12]
```diff
@@ -6516,13 +6516,81 @@ bool Item_func_regex::val_bool()
...
+}
+
+
 bool
 Item_func_regexp_instr::fix_length_and_dec(THD *thd)
 {
   if (agg_arg_charsets_for_comparison(cmp_collation, args, 2))
     return TRUE;
 
   re.init(cmp_collation.collation, 0);
```
Comment (gkodinov) [inline sql/item_cmpfunc.cc]: why keep the other init()?

File: sql/item_cmpfunc.cc [hunk 13]
```diff
@@ -6516,13 +6516,81 @@ bool Item_func_regex::val_bool()
 }
 
 
+// false = OK, true = error
+bool Regexp_processor_pcre::init(CHARSET_INFO *data_charset,
```
Comment (gkodinov) [inline sql/item_cmpfunc.cc]: Just pass the DTcollation parameter. data_charset is collation.collation. Or don't pass anything. cmp_collation is a member of the same class your method is a member of.

File: sql/item_cmpfunc.cc [hunk 14]
```diff
@@ -6514,28 +6514,203 @@ bool Item_func_regex::val_bool()
...
+                                           subject_ptr + match_byte_start) + 1;
+      else
+        return (longlong) lib_cs->numchars(subject_ptr,
+                                           subject_ptr + match_byte_end) + 1;
+    }
+
+    size_t next_offset= re.subpattern_end(0);
+    if (next_offset == byte_offset)
+      next_offset++;
+    byte_offset= next_offset;
```
[the remark this answers is not in the document]
Comment (gkodinov) [reply sql/item_cmpfunc.cc:6733]: this was already covered by the previous if().

**Review by gkodinov — APPROVED** [review]
LGTM. Please keep working with the final reviewer.

File: sql/item_cmpfunc.cc (hunk 12 above)
Comment (MohamedM216) [reply sql/item_cmpfunc.cc]: Although I knew it caused double initialization, I preferred to save a few of lines of code (fallback else blocks in case of null or non-const match_type) rather than prevent double initialization :)

Comment (gkodinov) [reply sql/item_cmpfunc.cc]: I'd advise against it. An API where you have to and can do init() twice is not a good one IMHO. But up to you.

**Review by gkodinov — APPROVED** [review]
LGTM. Please keep working with the final reviewer.

File: sql/item_cmpfunc.cc (hunk 12 above)
Comment (MohamedM216) [reply sql/item_cmpfunc.cc]: Thanks for the feedback. I already updated it to remove double initialization.

**Comment by MohamedM216** [comment]
@sanja-byelkin @gkodinov Hi, just following up on this PR. Thanks!

File: sql/item_cmpfunc.cc [hunk 15]
```diff
@@ -6531,13 +6604,136 @@ Item_func_regexp_instr::fix_length_and_dec(THD *thd)
...
+  longlong occurrence= 1;
+  if (arg_count > 3)
+  {
+    occurrence= args[3]->val_int();
+    if ((null_value= args[3]->null_value))
+      return 0;
+    if (occurrence <= 0)
+    {
+      my_error(ER_WRONG_PARAMETERS_TO_NATIVE_FCT, MYF(0), func_name());
+      return (null_value= true, 0);
```
Comment (sanja-byelkin) [inline sql/item_cmpfunc.cc:6639]: Now it can return NULL in any case but Item_func_regexp_instr::fix_length_and_dec do not sert maybe_null unconditionally (usual calculation for the functions upon the arguments) or at least set it for arguments number more than 3

File: sql/item_cmpfunc.cc [hunk 16]
```diff
@@ -6531,13 +6604,136 @@ Item_func_regexp_instr::fix_length_and_dec(THD *thd)
...
+
+      if (return_option == 0)
+        return (longlong) lib_cs->numchars(subject_ptr,
+                                           subject_ptr + match_byte_start) + 1;
+      else
+        return (longlong) lib_cs->numchars(subject_ptr,
+                                           subject_ptr + match_byte_end) + 1;
+    }
+
+    size_t next_offset= re.subpattern_end(0);
```
Comment (sanja-byelkin) [inline sql/item_cmpfunc.cc:6719]: Epty string matches processed incorrectly here:

SELECT REGEXP_INSTR('abc','(?=c)',1,1);
REGEXP_INSTR('abc','(?=c)',1,1)
3
SELECT REGEXP_INSTR('abc','(?=c)',1,2);
REGEXP_INSTR('abc','(?=c)',1,2)
3

(there is no second occurance here)

File: sql/item_cmpfunc.cc [hunk 17]
```diff
@@ -6516,13 +6516,86 @@ bool Item_func_regex::val_bool()
...
+      default:
+        my_error(ER_WRONG_VALUE_FOR_TYPE, MYF(0),
+                "match_type",
+                const_cast<String*>(match_type_str)->c_ptr_safe(),
+                "regexp_instr");
+        return true;
+      }
+    }
+  }
+  m_library_flags&= ~(PCRE2_CASELESS | PCRE2_MULTILINE | PCRE2_DOTALL);
```
Comment (sanja-byelkin) [inline sql/item_cmpfunc.cc:6561]: Have  default_regex_flags taken into considiration here?

File: mysql-test/main/regexp_instr_mysql8.test [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+SELECT REGEXP_INSTR(NULL,  'a');
+SELECT REGEXP_INSTR('abc', NULL);
+SELECT REGEXP_INSTR('abc', 'a', NULL);
+SELECT REGEXP_INSTR('abc', 'a', 1, NULL);
+SELECT REGEXP_INSTR('abc', 'a', 1, 1, NULL);
+SELECT REGEXP_INSTR('abc', 'a', 1, 1, 0, NULL);
+
+--echo # 8. Edge cases
+
+# Zero-length match
```
Comment (sanja-byelkin) [inline mysql-test/main/regexp_instr_mysql8.test:89]: where --echo disappered as it was above (please add to have references in the output file)

**Review by sanja-byelkin — CHANGES_REQUESTED** [review]
First of all is it MDEV-39520 or MDEV-39106 (PR title, commit comment and test headers contradicts)

Also there are otherer problem see below (I am no sure I have found all but for now it is enough to fix/answer first IMHO)

**Comment by MohamedM216** [comment]
> First of all is it MDEV-39520 or MDEV-39106 (PR title, commit comment and test headers contradicts)

It's MDEV-39520. I've updated the commit message. Sorry for confusion.

File: sql/item_cmpfunc.cc (hunk 17 above)
Comment (MohamedM216) [reply sql/item_cmpfunc.cc:6565]: The MySQL 8.0 behavior is that `match_type` fully overrides the default flags for the flags it can control. So the three flags controlled by `match_type` (CASELESS, MULTILINE, DOTALL) are cleared and re-set from the `match_type` string. Other flags from `default_regex_flags()` are intentionally preserved. If there's other intent, I can adjust.

File: mysql-test/main/regexp_instr_mysql8.test (hunk 18 above)
Comment (MohamedM216) [reply mysql-test/main/regexp_instr_mysql8.test]: OK, I'll add it.

File: sql/item_cmpfunc.cc (hunk 16 above)
Comment (MohamedM216) [reply sql/item_cmpfunc.cc]: Fixed. See new test cases added here https://github.com/MariaDB/server/pull/5035/changes#diff-af012131afb882730f2a3c60526c4211ffdeac4b2e484495bbf6b982d6f0f700R192

see code here https://github.com/MariaDB/server/pull/5035/changes#diff-1d0eeab9c76ca431af1f61a23ffd0cb93f4a38709e16005c0aa636ec9a6c313dR6722

File: sql/item_cmpfunc.cc (hunk 15 above)
Comment (MohamedM216) [reply sql/item_cmpfunc.cc:6655]: Fixed in fix_length_and_dec now, see https://github.com/MariaDB/server/pull/5035/changes#diff-1d0eeab9c76ca431af1f61a23ffd0cb93f4a38709e16005c0aa636ec9a6c313dR6604

**Comment by MohamedM216** [comment]
@sanja-byelkin Thanks for the detailed feedback. I've pushed a new patch. Looking forward to your review. Thanks!

**Comment by MohamedM216** [comment]
https://buildbot.mariadb.org/#builders/719/builds/18033 has environment-specific encoding issue. It passes locally and on other buidbots. Also the only part that fails is the Multibyte characters.
